### PR TITLE
ImageOverrider Implementation

### PR DESCRIPTION
**Pull request by XiShanYongYe-Chang**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

ImageOverrider  implementation

**Which issue(s) this PR fixes**:
Fixes https://github.com/karmada-io/karmada/issues/313

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: pkg/util/overridemanager/imageoverridemanager.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
+package overridemanager
```
Comment (RainbowMango) [inline pkg/util/overridemanager/imageoverridemanager.go:1]: I'd prefer to rename the file to `imageoverrider.go`.

**Comment by RainbowMango** [comment]
cc me when ready for review.

File: pkg/util/overridemanager/imageoverridemanager.go (hunk 1 above)
Comment (XiShanYongYe-Chang) [reply pkg/util/overridemanager/imageoverridemanager.go]: done.

**Comment by XiShanYongYe-Chang** [comment]
I've changed a lot, can you have a look, thanks. @RainbowMango

**Review by RainbowMango — CHANGES_REQUESTED** [review]

File: pkg/util/overridemanager/overridemanager.go [hunk 2]
```diff
@@ -207,40 +203,40 @@ func (o *overrideManagerImpl) getMatchingOverridePolicies(policies []policyv1alp
...
+	return overrideoption.ApplyJSONPatch(rawObj, parseJSONPatchesByPlaintext(overriders.Plaintext))
+}
 
-	objectJSONBytes, err := obj.MarshalJSON()
-	if err != nil {
-		return err
-	}
+func applyJSONPatchesByImageOverriders(rawObj *unstructured.Unstructured, imageOverriders []policyv1alpha1.ImageOverrider) error {
+	for index := range imageOverriders {
+		patches, err := imageoverrider.ParseJSONPatchesByImageOverrider(rawObj, &imageOverriders[index])
```
[the remark this answers is not in the document]
Comment (XiShanYongYe-Chang) [reply pkg/util/overridemanager/overridemanager.go:218]: I rename method ParseJSONPatchesByImageOverrider to BuildPatches.

File: pkg/util/overridemanager/imageoverrider/imageoverride.go [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+	switch rawObj.GetKind() {
+	case util.PodKind:
+		return parseJSONPatchesWithSpecContainersPath("spec/containers", rawObj, imageOverrider)
+	case util.ReplicaSetKind:
+		return parseJSONPatchesWithSpecContainersPath("spec/template/spec/containers", rawObj, imageOverrider)
+	case util.DeploymentKind:
+		return parseJSONPatchesWithSpecContainersPath("spec/template/spec/containers", rawObj, imageOverrider)
+	case util.StatefulSetKind:
+		return parseJSONPatchesWithSpecContainersPath("spec/template/spec/containers", rawObj, imageOverrider)
+	}
```
[the remark this answers is not in the document]
Comment (XiShanYongYe-Chang) [reply pkg/util/overridemanager/imageoverrider/imageoverride.go:40]: Done.

File: pkg/util/overridemanager/overridemanager.go [hunk 4]
```diff
@@ -2,17 +2,17 @@ package overridemanager
...
-	jsonpatch "github.com/evanphx/json-patch/v5"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/klog/v2"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	clusterv1alpha1 "github.com/karmada-io/karmada/pkg/apis/cluster/v1alpha1"
 	policyv1alpha1 "github.com/karmada-io/karmada/pkg/apis/policy/v1alpha1"
 	"github.com/karmada-io/karmada/pkg/util"
+	"github.com/karmada-io/karmada/pkg/util/overridemanager/imageoverrider"
+	"github.com/karmada-io/karmada/pkg/util/overridemanager/overrideoption"
```
[the remark this answers is not in the document]
Comment (XiShanYongYe-Chang) [reply pkg/util/overridemanager/overridemanager.go:15]: Sorry. I misunderstood you earlier. I'll change it back.

**Pull request merged**